python/cudf_polars/cudf_polars/experimental/rapidsmpf/nodes.py

```diff
@@ -0,0 +1,436 @@
+                for ch_idx in range(n_children)
+            ):
+                # Ready to produce the output chunk for seq_num.
+                # Evaluate and send.
+                df = await asyncio.to_thread(
+                    ir.do_evaluate,
+                    *ir._non_child_args,
+                    *[
+                        (
+                            staged_chunks[ch_idx][0]
+                            if ch_idx in bcast_indices
+                            else staged_chunks[ch_idx].pop(seq_num)
+                        )
+                        for ch_idx in range(n_children)
+                    ],
+                )
+                await ch_out.data.send(
+                    ctx,
+                    Message(
+                        TableChunk.from_pylibcudf_table(
+                            seq_num,
+                            df.table,
+                            DEFAULT_STREAM,
+                            exclusive_view=True,
+                        )
+                    ),
+                )
+                seq_num += 1
+            elif not accepting_data:
+                if any(
+                    staged_chunks[ch_idx]
+                    for ch_idx in range(n_children)
+                    if ch_idx not in bcast_indices
+                ):
+                    raise RuntimeError(
+                        f"Leftover data in staged chunks: {staged_chunks}."
+                    )
+                break  # All channels have finished
+
+        await ch_out.data.drain(ctx)
+
+
+@define_py_node()
+async def multicast_node_bounded(
+    ctx: Context,
+    ch_in: ChannelPair,
+    *chs_out: ChannelPair,
+) -> None:
+    """
+    Bounded multicast node for rapidsmpf.
+
+    Each chunk is broadcasted to all output channels
+    as it arrives.
+
+    Parameters
+    ----------
+    ctx
+        The context.
+    ch_in
+        The input ChannelPair.
+    chs_out
+        The output ChannelPairs.
+    """
+    # TODO: Use multiple streams
+    async with shutdown_on_error(ctx, ch_in.data, *[ch.data for ch in chs_out]):
+        while (msg := await ch_in.data.recv(ctx)) is not None:
+            table_chunk = TableChunk.from_message(msg)
+            for ch_out in chs_out:
+                await ch_out.data.send(
+                    ctx,
+                    Message(
+                        TableChunk.from_pylibcudf_table(
+                            table_chunk.sequence_number,
+                            table_chunk.table_view(),
+                            table_chunk.stream,
+                            exclusive_view=False,
+                        )
+                    ),
+                )
+
+        await asyncio.gather(*(ch.data.drain(ctx) for ch in chs_out))
+
+
+async def _send_table_chunks(
+    ctx: Context,
+    table_chunks: list[TableChunk],
+    ch_out: Channel,
+) -> None:
+    """Send a list of table chunks to the output channel."""
+    async with shutdown_on_error(ctx, ch_out):
```

Comment:
Out of curiosity, when do we need this `shutdown_on_error`? Is having the `with shutdown_on_error` in the caller (`multicast_node_unbounded`) sufficient?